### ISSUE TITLE
what kinds of text do I need to draw the lables around the 3D axis

### ISSUE THREAD

**Issue by asmwarrior** [markaren/threepp#216]
Hi, I would like to draw some 3D math graph like the one below:

https://stemkoski.github.io/MathBox/graph3d-surface.html

So, I need some text types. I looked at the sample code, and I see that `Text2D` class is mainly used for HUD. But for me, I would like to give the text label 3 coordinates `x,y,z`, so that it can rotate around the 3D scene.

Any suggestions? Thanks.

**Comment by asmwarrior** [comment]
Thanks for the help, let me try it. I come from other OpenGL tutorial, in that world, the `SpriteMaterial` mainly be called `billboard` label. Because the text surface is always perpendicular to the camera's axis.

**Comment by asmwarrior** [comment]
~~~~
diff --git a/examples/extras/core/fonts.cpp b/examples/extras/core/fonts.cpp
index fedba92..8851c65 100644
--- a/examples/extras/core/fonts.cpp
+++ b/examples/extras/core/fonts.cpp
@@ -119,13 +119,17 @@ int main() {
 
     if (font) {
         const auto material = MeshPhongMaterial::create();
+        const auto material1 = SpriteMaterial::create();
         material->side = Side::Double;
         material->color = Color::orange;
+        
+        material1->side = Side::Double;
+        material1->color = Color::green;
 
         textMesh3d = Text3D::create(ExtrudeTextGeometry::Options(*font, textSize, 1), displayText, material);
         textMesh3d->castShadow = true;
 
-        textMesh2d = Text2D::create(TextGeometry::Options(*font, textSize), displayText, material);
+        textMesh2d = Text2D::create(TextGeometry::Options(*font, textSize), displayText, material1);
         textMesh2d->position.z = 2;
 
         textMesh3d->geometry()->center();
~~~~

Some good news: with the above patch applied to the examples/fonts.cpp file, I can keep the Text2D object like a billboard.

See the gif below when I rotate the scene.
![font-billboard](https://github.com/markaren/threepp/assets/561818/96cc165a-07f1-4e6a-9c04-1c74780e0810)

Now, I need to create a lot of such objects(`textMesh2d`), and draw them around an axis.

To change the position, I need to set the `textMesh2d->position.z`, `textMesh2d->position.x`, and `textMesh2d->position.y`.

**Comment by asmwarrior** [comment]
I have try to add two billboard text labels in my wxWidgets port, see this commit:

[add two billboard text lables in the scene](https://github.com/asmwarrior/threepp/commit/9f167b19858f988e26145065957225bdca798021)

The result is it does have two lables, but the text is too large, I have already set the variable `unsigned int textSize = 1;`, see the image shot below:

![image](https://github.com/markaren/threepp/assets/561818/2ba36764-dd4f-44b5-8a10-cd035aa87884)


Any ideas?

Another issue I see is that my added two billboard text has different size, the text near the camera is bigger than the other one.
I think it should be the same size, I mean the text should be the same size. I'm not sure how the shader works, but it should remove the distance factor of the perspective projection.

Thanks.

**Comment by markaren** [comment]
1. Scale the geometry or mesh
2. Might work to use an ortographic projection in a separate render call like for the HUD

**Comment by asmwarrior** [comment]
> 1. Scale the geometry or mesh
> 
>     2. Might work to use an ortographic projection in a separate render call like for the HUD
> 
> 
> Disclaimer. I am in no way an expert on this. You might want to search three.js for a solution

Thanks, I think the method 2 is the direction. I will try to see what I can do. I know nothing about javascript, so reading the three.js may need more effert.

**Comment by asmwarrior** [comment]
Something like the below links:

https://codepen.io/sureshwisdom/pen/gObavym

I need some time to learn the logic. Because the shader is hidden inside the threepp framework, so I need to debug (step into) its source code.

**Comment by markaren** [comment]
It's as easy as setting `sizeAttenuation=false` on the `SpriteMaterial`
https://threejs.org/docs/#api/en/materials/SpriteMaterial.sizeAttenuation

**Comment by asmwarrior** [comment]
> It's as easy as setting `sizeAttenuation=false` on the `SpriteMaterial` https://threejs.org/docs/#api/en/materials/SpriteMaterial.sizeAttenuation
> 

Thanks, setting the `sizeAttenuation=false` make the text labels show in the same size. But the font is extreme large. The option: `textSize` could only be `unsigned int`, but even I set this value to `1`, the font is still big. Could this value be a `float` type?

~~~~
textMesh2dArray.push_back(Text2D::create(TextGeometry::Options(font2, textSize), displayText, material1));
~~~~

**Comment by asmwarrior** [comment]
> Something like the below links:
> 
> https://codepen.io/sureshwisdom/pen/gObavym
> 
> I need some time to learn the logic. Because the shader is hidden inside the threepp framework, so I need to debug (step into) its source code.

I read the code in the above link, I see such lines:

~~~~
        var sprite = new THREE.Sprite( spriteMaterial );
        sprite.scale.set(0.5 * fontsize, 0.25 * fontsize, 0.75 * fontsize);
        return sprite;  
~~~~

If I tweak the `scale`, I got the different text view size. So, my question is: `SpriteMaterial` has a scale option to let me control the text size?

Thanks.

**Comment by asmwarrior** [comment]
OK, I may find a solution with such patch.

~~~~
diff --git a/cb/main.cpp b/cb/main.cpp
index 22ae569..7d0278d 100644
--- a/cb/main.cpp
+++ b/cb/main.cpp
@@ -262,6 +262,7 @@ bool OpenGLCanvas::InitializeOpenGL()
     const auto material1 = SpriteMaterial::create();
     material1->side = Side::Double;
     material1->color = Color::green;
+    material1->sizeAttenuation = false;
 
 
     textMesh2dArray.push_back(Text2D::create(TextGeometry::Options(font2, textSize), displayText, material1));
@@ -270,6 +271,10 @@ bool OpenGLCanvas::InitializeOpenGL()
     textMesh2dArray[0]->position.z = 5;
     textMesh2dArray[1]->position.z = -5;
 
+    textMesh2dArray[0]->scale.set(0.05,0.05,0.05);
+    textMesh2dArray[1]->scale.set(0.05,0.05,0.05);
+
+
     textMesh2dArray[0]->geometry()->center();
     textMesh2dArray[1]->geometry()->center();
 
~~~~

Here is the result image shot:

![image](https://github.com/markaren/threepp/assets/561818/4ed09be1-ac28-4383-bbb1-a7789485ee51)

**Comment by markaren** [comment]
> Could this value be a float type

Yeah, no you are right, this could be a float. Will change it.

**Comment by asmwarrior** [comment]
> > Could this value be a float type
> 
> Yeah, no you are right, this could be a float. Will change it.
> 
> See [4d29de3](https://github.com/markaren/threepp/commit/4d29de34073d33ae48151ee7546be62622e0d4e6)

I think you change to the code base has bugs.

If you try to run the `fonts` example, and set the `textSize` to `0.5` (the original value is `10`), you will see the 2D text is missing. the 3D text is still there.

**Comment by markaren** [comment]
Fixed (9b99e09ce465608e1bf43e2539b09a007543a07d)

**Comment by asmwarrior** [comment]
> Fixed ([9b99e09](https://github.com/markaren/threepp/commit/9b99e09ce465608e1bf43e2539b09a007543a07d))

Thanks, it works OK now.
